File: app/src/main/java/com/example/anchornotes/data/db/NoteSearchDao.java
package com.example.anchornotes.data.db;

import androidx.room.Dao;
import androidx.room.RawQuery;
import androidx.sqlite.db.SupportSQLiteQuery;
import java.util.List;

@Dao
public interface NoteSearchDao {
    @RawQuery(observedEntities = {NoteEntity.class})
    List<NoteEntity> search(SupportSQLiteQuery query);
}

File: app/src/main/java/com/example/anchornotes/data/db/NoteSearchDao.kt
package com.example.anchornotes.data.db

import androidx.lifecycle.LiveData
import androidx.room.*
import androidx.sqlite.db.SupportSQLiteQuery

/**
 * DAO interface for search operations.
 * Uses @RawQuery for dynamic query building based on filter parameters.
 */
@Dao
interface NoteSearchDao {
    
    /**
     * Search notes with filters.
     * Uses a raw query approach for maximum flexibility with dynamic filters.
     * 
     * Note: This implementation uses "ANY tag" matching - a note matches
     * if it has ANY of the selected tags (OR logic), not ALL tags.
     */
    @RawQuery(observedEntities = [NoteEntity::class])
    fun searchNotes(query: SupportSQLiteQuery): LiveData<List<NoteEntity>>
}

